model/reliability/db.go

```diff
@@ -26,8 +26,72 @@ import (
 	"go.mongodb.org/mongo-driver/bson"
 )
 
+// DateBoundaries returns the date boundaries when splitting the period between 'start' and 'end' in groups of 'numDays' days.
+// The boundaries are the start dates of the periods of 'numDays' (or less for the last period), starting with 'start'.
+func (filter TaskReliabilityFilter) DateBoundaries() []time.Time {
+	start := filter.AfterDate
+	end := filter.BeforeDate
+	numDays := filter.GroupNumDays
+
+	if numDays <= 0 {
+		numDays = 1
+	}
+
+	start = util.GetUTCDay(start)
+	end = util.GetUTCDay(end)
+
+	boundaries := []time.Time{}
+	duration := 24 * time.Hour * time.Duration(numDays)
+	duration = -1 * duration
+	boundary := end.Add(24 * time.Hour)
+
+	for boundary.After(start) {
+		boundaries = append(boundaries, boundary)
+		boundary = boundary.Add(duration)
+	}
+	boundaries = append(boundaries, boundary)
+	return boundaries
+}
+
+// BuildTaskPaginationOrBranches builds an expression for the conditions imposed by the filter StartAt field.
+func (filter TaskReliabilityFilter) BuildTaskPaginationOrBranches() []bson.M {
```

Comment:
I'd be inclined to make the public functions return concrete options rather than queries/pipelines.

Reply:
Do you mean that the public method  should returns the array of TaskReliability structs?
 
The code is laid out as it it as I was copying the way stats/db.go was implemented.

(I also made some of the functions public so that they could unit tested easier).

Reply:
you should be able to do all your unittesting in the same package (e.g. reliability) without needing to make things public, but I might have missed something in the tests.

Evergreen has always just made too much public, particularly in the models package (and sub-packages), where there was an early attempt to split up a big package into multiple packages. This attempt (which gives us the structure we have today,) was half completed, and the result isn't particularly compelling (to me.) Because the business objects that models represent are rarely orthogonal so there ends up being a lot of (non-obvious) dependencies between sub-model packages with a lot more public functions, for the "benefit" of shorter names for types and a lot more boilerplate. 

It is the way it is, and we do have an intention to address it in the medium term (mostly because it's quite hard to back out of), but it does mean that the patterns in the models hierarchy are difficult to copy, or learn from.

Reply:
I think the push to make methods/functions in the model packages more _functional_ is generally a good one. 

As the author of the machinery beneath the db.Aggregate function, I'm also deeply suspicious. I think I changed the stats package to call out to the new driver's client directly, which might give you more options. I don't think db.Aggregate is entirely broken (and both use the new driver.)

Reply:
That seems reasonable to me. 

>  I think I changed the stats package to call out to the new driver's client directly, which might give you more options

Where is that in the code base?

It would be interesting to have direct access to aggregation calls. It would allow the pagination to be more efficient.

I don't think it is currently worth the effort, given these aggregations are relativey efficient (and the limit / page size generally means that pagination isn't required).

Having looked at the wrapped aggregate functions, I appreciate the simplicity.